### ISSUE TITLE
Add a PHP binding (litsea-php, ext-php-rs)

### ISSUE THREAD

**Issue by mosuka** [mosuka/litsea#204]
Part of #200. Depends on #201 (`litsea-binding-core`).

## Background

PHP applications (search indexers, CMS text pipelines) currently cannot call Litsea without invoking the CLI per request. Lindera ships [`lindera-php`](https://github.com/lindera/lindera/tree/main/lindera-php) as a workspace member built with [`ext-php-rs`](https://github.com/davidcole1340/ext-php-rs) (0.15.x), packaged with a `composer.json` of `"type": "php-ext"` requiring PHP >= 8.1, and tested with PHPUnit plus php-cs-fixer for style.

Note that PHP is the one target where a "just install it" story does not exist the way it does for PyPI/npm/RubyGems: a PHP extension is a shared object that has to be built for the target PHP ABI and enabled in `php.ini`. The issue therefore has to cover the build/install documentation explicitly.

## Goal

Add a `litsea-php` workspace member exposing Litsea to PHP 8.1+ via ext-php-rs, distributed as source through Packagist (`php-ext`) with documented build steps (no embedded models, per #200).

Proposed API:

```php
<?php
use Litsea\Segmenter;
use Litsea\Extractor;
use Litsea\Trainer;
use Litsea\Language;
use Litsea\Upos;

$seg = Segmenter::fromPath(Language::Japanese, 'models/japanese.model');
$seg->segment('すもももももももものうち');
// ['すもも', 'も', 'もも', 'も', 'もも', 'の', 'うち']
$seg->segmentBatch(['...', '...']);

$pos = Segmenter::fromPath(Language::Japanese, 'models/japanese_pos.model');
$pos->segmentWithPos('すもももももももものうち');
// [['surface' => 'すもも', 'pos' => 'NOUN'], ...]

Segmenter::fromBytes(Language::Korean, file_get_contents('korean.model'));
Segmenter::fromUri(Language::Chinese, 'https://example.com/chinese.model');

(new Extractor(Language::Japanese))->extract('corpus.txt', 'features.txt');
(new Trainer(0.01, 10000, 'features.txt'))->train('japanese.model');
```

Design points:

- `[lib] name = "litsea_php"`, `crate-type = ["cdylib", "lib"]`, `build.rs` from ext-php-rs.
- `Language` / `Upos`: decide between PHP 8.1 native `enum` (if ext-php-rs can register one in the target version) and class constants + string accessors; document the choice. Strings must round-trip with the Rust canonical names.
- POS results as associative arrays (`surface` / `pos`) or a small `Litsea\Token` class — pick one and keep it consistent with the other bindings' token shape from `litsea-binding-core`.
- `LitseaError` mapped to a `Litsea\LitseaException` hierarchy extending `\RuntimeException`.
- The async URI loader has to be driven from a blocking runtime (PHP has no event loop here) — use the strategy decided in `litsea-binding-core`.
- Long-running training under a web SAPI is a foot-gun; document that training is intended for CLI SAPI usage and honour `max_execution_time` / signals where possible via the `&AtomicBool` cancellation flag.
- Document the install path: `cargo build --release`, copy `liblitsea_php.so` / `.dylib` / `.dll` into the extension dir, add `extension=litsea_php` to `php.ini`; mention `cargo php install` from ext-php-rs as the convenience route.

## Task list

- [ ] Create `litsea-php/` (`Cargo.toml`, `build.rs`, `composer.json`, `src/`) and add it to the root workspace
- [ ] `src/lib.rs`, `segmenter.rs`, `extractor.rs`, `trainer.rs`, `language.rs`, `upos.rs`, `convert.rs`, `error.rs`
- [ ] `fromPath` / `fromBytes` / `fromUri` constructors
- [ ] `segment`, `segmentBatch`, `segmentWithPos`, `segmentWithPosBatch`
- [ ] Extractor / Trainer classes with cancellation
- [ ] Decide the `Language` / `Upos` representation (native enum vs class constants) and document it
- [ ] PHPUnit tests: parity per language, POS output, error cases, training round-trip
- [ ] `examples/` and php-cs-fixer config
- [ ] `README.md` + `README_ja.md` including the build/`php.ini` install instructions; EN/JA docs page under `docs/src` / `docs/ja/src` and `SUMMARY.md` entries
- [ ] CI: test job (setup PHP 8.1+, build extension, run PHPUnit) in `regression.yml` and `release.yml`; Packagist release notes in `release.yml`

## Acceptance criteria

- With the extension loaded, `Segmenter::fromPath(...)->segment(...)` matches `litsea segment` output for Japanese, Chinese, Korean, and English.
- `segmentWithPos` on a `*_pos.model` matches `litsea segment --pos`.
- Error paths throw `Litsea\LitseaException` rather than crashing the PHP process.
- Training from PHP produces a model the Rust CLI can load.
- PHPUnit and php-cs-fixer pass in CI; `cargo clippy --workspace --all-targets -- -D warnings` is clean.
- The Composer package contains no `*.model` file.

**Comment by mosuka** [comment]
## Investigation report

Builds on #201 (shared core), #202 (Python), and #203 (Node.js). What is specific to PHP:

### ext-php-rs

Version 0.15.15, the same line Lindera uses; its CI runs PHP 8.5, and the local environment here is PHP 8.5.4 with `php-config` and libclang-18, so the toolchain is available. Patterns confirmed from [`lindera-php`](https://github.com/lindera/lindera/tree/main/lindera-php) and the ext-php-rs guide:

- `#[php_class]` + `#[php(name = "Litsea\\Segmenter")]` for namespaced classes, `#[php_impl]` with `pub fn __construct() -> PhpResult<Self>`, and `#[php_module] pub fn get_module(module: ModuleBuilder) -> ModuleBuilder` registering each class.
- **Methods are renamed to camelCase by default** (`change_method_case` overrides it), so `segment_with_pos` reaches PHP as `segmentWithPos` without per-method annotations.
- **Custom exception classes are supported**: `#[php(extends(ce = ce::exception, stub = "\\Exception"))]` for a base deriving from `\Exception`, and `#[php(extends(MyBase))]` for Rust-defined parents. So the Python binding's hierarchy can be mirrored exactly, rather than falling back to a generic exception the way `lindera-php` does.
- PHP objects are not `Send`/`Sync`; interior mutability goes through `RefCell`, as `lindera-php` does.

### Concurrency: PHP is the exception among the bindings

Python releases the GIL (#202) and Node.js moves the work to libuv's threadpool (#203) — both of which make `CancelToken` usable *while* training runs. **Neither is available here.** A PHP request is single-threaded, and `pcntl` signal handlers cannot interrupt a blocking native call, so nothing in PHP can run while `train()` is executing.

The honest consequence, which the docs and tests must state rather than paper over: in PHP a `CancelToken` can only be set **before** `train()` is called (for example from state carried across a CLI loop). Cancelling still behaves as everywhere else — training stops at its next check point and the partial model is written — but there is no in-flight cancellation. This is a property of the host, not a gap in the binding, and it is the reason #203's event-loop test has no PHP analogue.

### Distribution

Unlike PyPI and npm, there is no "install a prebuilt binary" story: a PHP extension is a shared object built against a specific PHP ABI and enabled in `php.ini`. The package on Packagist is therefore source plus documentation (`"type": "php-ext"`, as `lindera-php` does), and the README has to carry the build and `php.ini` steps. `composer.json` requires PHP >= 8.1; testing uses PHPUnit.

**Comment by mosuka** [comment]
## Implementation plan

### Layout

```text
litsea-php/
├── Cargo.toml     # [lib] name = "litsea_php", crate-type = ["cdylib", "lib"], ext-php-rs 0.15
├── build.rs
├── composer.json  # "type": "php-ext", php >= 8.1, phpunit
├── src/           # lib.rs, segmenter.rs, trainer.rs, token.rs, metrics.rs, error.rs
├── tests/         # PHPUnit
└── examples/
```

### PHP API

```php
use Litsea\Segmenter;
use Litsea\Extractor;
use Litsea\Trainer;
use Litsea\CancelToken;

$seg = Segmenter::open('japanese', 'models/japanese.model');
$seg->segment('これはテストです。');        // ['これ', 'は', 'テスト', 'です', '。']
$seg->segmentBatch([...]);
$seg->hasPos();                             // the model kind is detected, no flag

$pos = Segmenter::open('ja', 'models/japanese_pos.model');
foreach ($pos->segmentWithPos('これはテストです。') as $token) {
    echo $token->surface, ' ', $token->pos, ' ', $token->start, '..', $token->end, "\n";
}

(new Extractor('japanese'))->extract('corpus.txt', 'features.txt');
$metrics = (new Trainer(0.01, 10000, 'features.txt'))->train('japanese.model');
```

Design decisions:

- **Exception hierarchy mirroring the Python binding**: `Litsea\LitseaException extends \Exception`, with `InvalidArgumentException`, `ModelException`, `IoException`, `ParseException`, `UnsupportedException`, and `PosUnavailableException` deriving from it. ext-php-rs supports this through `#[php(extends(...))]`, so there is no reason to fall back on generic exceptions.
- **Methods keep their default camelCase rename**, which matches PHP conventions.
- **`Token` is a class with public properties** (`surface`, `pos`, `start`, `end`); `pos` is `null` for untagged tokens. Byte offsets, as in every binding — PHP strings are byte strings, so `substr($text, $token->start, $token->end - $token->start)` works directly, unlike JavaScript.
- **Cancellation is documented as pre-call only.** `train()` accepts a `CancelToken`, and cancelling before the call still yields the partial model; there is no in-flight cancellation, because a PHP request cannot run code while a native call blocks. The docs and README say so explicitly, and a test covers the pre-call path.
- Everything blocks the request, which is fine for CLI usage and hostile under a web SAPI; the README says to train from the CLI SAPI.

### Repository changes

- Root `Cargo.toml`: new member (`default-members` already excludes bindings).
- `regression.yml` / `release.yml`: a `test-litsea-php` job (`shivammathur/setup-php` → build the extension → PHPUnit with `-d extension=...`). No publish job: Packagist consumes the git tag, so releasing needs no credentials — unlike PyPI (#202) and npm (#203).
- `Makefile`: `test-litsea-php`, `build-litsea-php`, `clean-litsea-php`, `lint-litsea-php`.
- `dependabot.yml`: a `composer` entry for `/litsea-php`.
- Docs: `docs/src/bindings/php.md` + the JA mirror, `SUMMARY.md` entries, status tables, README row.

### Verification

1. Parity against `litsea segment` for all four languages, comparing the CLI's rendered line (as in #202 and #203).
2. POS parity against `litsea segment --pos`.
3. Byte offsets: `substr()` on the input reproduces each surface.
4. Errors: each failure throws the matching `Litsea\*Exception`, and `catch (LitseaException)` catches them all.
5. Training round-trip: extract → train from PHP, then load the model with the CLI.
6. Cancellation before `train()` returns partial metrics and still writes the model.
7. CI parity: `cargo fmt --check`, `cargo clippy --workspace --all-targets -D warnings`, `cargo test --all-features`, `markdownlint-cli2`, both mdbook builds.

**Comment by mosuka** [comment]
## Implementation report

`litsea-php` is implemented on `feat/litsea-php` (34 files, +3230 / -12).

### What landed

- **`Litsea\Segmenter`** — `open` / `fromBytes` / `fromUri`, `segment`, `segmentBatch`, `segmentTokens`, `segmentWithPos`, `segmentWithPosBatch`, `language()`, `hasPos()`
- **Training** — `Litsea\Extractor`, `Trainer`, `PerceptronTrainer`, `TwoStageTrainer`, `CancelToken`, plus metric classes
- **Exceptions** — `Litsea\LitseaException` with one subclass per error kind, matching the Python binding
- **Packaging** — `composer.json` (`"type": "php-ext"`), PHPUnit config, build and `php.ini` instructions in the README

### Verification

32 PHPUnit tests (193 assertions) and 4 Rust unit tests pass on PHP 8.5.4. Parity is measured against the `litsea` CLI for all four languages, comparing its rendered output line (same approach as #202 and #203).

Sabotage check — removing `#[php(extends(LitseaException))]` from one exception class:

```text
1) Litsea\Tests\SegmenterTest::testEveryExceptionDerivesFromTheBase
Litsea\PosUnavailableException
Failed asserting that false is true.
```

The hierarchy test fails, as it should.

Also green locally: `cargo fmt --check`, `cargo clippy --workspace --all-targets -- -D warnings`, `cargo test --all-features`, `make test-litsea-php`, `markdownlint-cli2` (110 files), both mdbook builds.

### Three things measurement corrected

1. **The extension registers under the crate name with its hyphen** — `extension_loaded('litsea-php')`, not `litsea_php`. The bootstrap's guard failed loudly on the first run and pointed straight at it.
2. **ext-php-rs renames properties as well as methods to camelCase**, so metrics read as `$metrics->numInstances`, not `num_instances`. Four tests failed until the assertions matched reality; the READMEs and docs use the measured names.
3. **`fromBytes` cannot take `Vec<u8>`** — ext-php-rs rejects a PHP string for that parameter ("Invalid value given for argument `data`"). Model files are UTF-8 text, so the argument is a PHP string, and the doc comment says why.

### The PHP-specific design decision

`CancelToken` is documented and tested as **pre-call only**. Python releases the GIL (#202) and Node.js moves training to libuv's threadpool (#203), so both can cancel a run in flight. A PHP request is single-threaded and `pcntl` handlers cannot interrupt a blocking native call, so no PHP code runs during `train()`. Rather than implying a capability that does not exist, the README, the docs page, and the module documentation all state the limitation, and the test covers the path that does work: cancelling beforehand stops training at its first check point and still writes the partial model.

There is also no `publish-php` job, and none is needed — Packagist consumes the git tag, so unlike #202 (`PYPI_TOKEN`) and #203 (`NPM_TOKEN`) this binding needs no credentials.

### Note for #205 / #206

Ruby's GVL is the next instance of the same question. `magnus` offers `without_gvl`, so Ruby should be able to do what Python does — but that needs measuring, not assuming, and the Python GIL test (assert that the canceller thread ran inside the training window) is the shape to reuse.
